File: leetcode/34.find-first-and-last-position-of-element-in-sorted-array.py
# @leet start
class Solution:
    def searchRange(self, nums: list[int], target: int) -> list[int]:
        # Do two binary searches. One to find the first occurance of
        # the target, and another to find the last occurance.
        n = len(nums)

        start = 0
        end = n - 1

        while start <= end:
            mid = (start + end) // 2

            if nums[mid] < target:
                start = mid + 1
            else:
                end = mid - 1

        if start == n or nums[start] != target:
            return [-1, -1]

        # Start is now equal to the index at which the first occurance
        # would occur, if it occurs
        first_position = start

        # Now find the last occurance
        start = first_position
        end = n - 1

        while start <= end:
            mid = (start + end) // 2

            if nums[mid] > target:
                end = mid - 1
            else:
                start = mid + 1

        second_position = end

        return [first_position, second_position]


# @leet end

File: leetcode/34.find-first-and-last-position-of-element-in-sorted-array.js
// @leet start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function (nums, target) {
  // Find first position of target
  let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);

    if (nums[mid] >= target) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  const firstIdx = left;

  if (firstIdx === nums.length || nums[left] !== target) {
    return [-1, -1];
  }

  // Find first position after target, of a non-target num
  left = firstIdx + 1;
  right = nums.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);

    if (nums[mid] > target) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return [firstIdx, left - 1];
};
// @leet end
